BestFloww/BestFlow
# Adding Intent Menu with search

## Pull Request Summary
Adds an Intent Menu to Analysis Page with the following features:
- Can open/close with sliding animation based on props from Analysis Page
- Lists all intents in the analyzed transcript in a scrollable box
- Clicking on any of the intents will properly direct you to the correct page (guarantees proper index)
- Has an input field where you can type in text to search the exact intent you want, which narrows down the list in realtime. It is not case-sensitive, but is otherwise sensitive to exact spelling, punctuation, etc.
- The part of the intent that matches the search string is highlighted in realtime as well
- Clicking anywhere outside of the Intent Menu closes it

What needs to be done from this point: there is a tentative hamburger button at the bottom of the page that opens the sidebar. Eventually this should be in the navbar that exists on some other branch (Kaylee's I think). Right now you can't use that hamburger button to close the sidebar, bc the same behavior will be expected from the hamburger button the navbar (it will be covered by the sidebar and you will instead use the X button to close, which exists in the sidebar already).

I am really proud of this PR! The code is fairly long but I felt that a lot of the logic to achieve the results I wanted was intuitive and many features worked on first try.

### Change types


### Testing

### How to manually test this PR (QA instructions)
![image](https://user-images.githubusercontent.com/69657338/202885849-a7e1bda5-d21d-4cc8-be90-d37ddeea4d2b.png)
Right now in the DB, I uploaded a large transcript with lots of intents with Project ID 3. This may no longer be true by the time you QA this ticket, but try using that one since it has the most intents. Notice how the intents change based on which one you click in the list. Type keywords into the input field and notice how the results are narrowed down.

## Questions

1. How's the styling and user experience? We can refine the search a little bit to be less strict/sensitive, but I don't know if it's worth the time.

**Outdated questions (answered!):**
1. I have not yet added testing bc a lot of is blocked by the way I coded the animation. Right now, the Intent Menu is always rendered and it simply moves on and off screen depending on the isOpen prop. Is this the right way to do it? Is there a better way? How do I test for whether the menu is in the document? I cannot find an expect function that tests for whether an element is onscreen - only if it's loaded in the DOM in which case it is always true.
2. I have added a "testing" skeleton outlining all the unit tests I want to do. Let me know if anything should be added or removed.

**Resources:**
- Detecting clicks outside a component: https://blog.logrocket.com/detect-click-outside-react-component-how-to/
- Simple animation: https://blog.avneesh.tech/create-an-animated-sidebar-with-tailwindcss-in-react
- New icons were added, see the icon biblio
